### ISSUE TITLE
Add csdp to the Coq Platform

### ISSUE THREAD

**Issue by palmskog** [coq/platform#39]
The [psatz tactic](https://coq.inria.fr/distrib/V8.12.0/refman/addendum/micromega.html#short-description-of-the-tactics) uses the external tool [csdp](https://github.com/coin-or/Csdp).

To make it convenient for users of any platform to call `psatz`, I propose that csdp is included in the Coq Platform. Debian and Ubuntu users can already install it as the package `coinor-csdp`.

**Comment by MSoegtropIMC** [comment]
@palmskog : there is a homebrew package for csdp, but neither a macports nor a cygwin package. So the easiest path might be to create an opam "from sources" package for csdp, but as far as I know coinor, this might get a bit involved. Is there someone who might be able to help with this?

In general I have some interest in a reasonable install method for coinor, since I also sometimes want it for other things, but I am not sure I have time for this.

I am not in favor of including it if we can't support it on half of the Macs and on Windows. Also I am not sure about the other Linux distros.

**Comment by palmskog** [comment]
Unfortunately I don't know anyone with the right expertise who could help out with packaging this. Could we maybe add some kind of "help wanted" tag here to the issue and postpone inclusion until we find more help? This is not a critical package for me, but several people have wondered about using `psatz` more easily.

**Comment by MSoegtropIMC** [comment]
Yes - we have a help wanted label.

As I said eventually I will likely do it if no one else does it, just because for many LP problems one needs a second or third opinion (see e.g. https://gist.github.com/MSoegtropIMC/2d4fdfe6b1e3e3955c63084523db3ca7) or a formally verified solver ... But it might take a while.

**Comment by erikmd** [comment]
I Cc @proux01 in case he'd have some experience running csdp on Windows…
(indeed, our ValidSDP library relies on his [`osdp` libary](https://cavale.enseeiht.fr/osdp/) which is an ocaml library providing a unified library for several SDP solvers, including but not limited to csdp)

**Comment by erikmd** [comment]
Also, see the [`conf-csdp`](https://opam.ocaml.org/packages/conf-csdp/) opam package (which is one of the dependencies of [`osdp`](https://opam.ocaml.org/packages/osdp/))

**Comment by MSoegtropIMC** [comment]
The opam package has just:
```
depexts: [
  ["coinor-csdp"] {os-family = "debian"}
  ["csdp"] {os-distribution = "centos"}
]
```
which is quite a bit below the range of platforms we support and at least on Windows and MacPorts it won't be just adding depext lines.

**Comment by proux01** [comment]
Indeed, the conf-csdp package only checks that csdp is installed but doesn't actually package it (which is not trivial since its C code depends on blas/lapack and fortran). And indeed, the depext packages availability seems rather limited.

**Comment by MSoegtropIMC** [comment]
I postponed this package inclusion to the next release which will likely happen in November (for Coq 8.14)

**Comment by palmskog** [comment]
As per discussion on Zulip, inclusion of csdp in the platform has been postoned further, possibly indefinitely.

**Comment by MSoegtropIMC** [comment]
@palmskog : as I said eventually I will likely do it, but right now there are too many things which give a higher benefit for less work (coq-hammer e.g.).

**Comment by palmskog** [comment]
I agree about coq-hammer, this was mainly just to inform that people can't start to rely on csdp in 2021.11.

**Comment by MSoegtropIMC** [comment]
Do you see a big benefit in it? I tried it a few times and it usually didn't do what I want, so I tend to use either lra or coq-interval or a combination of both. This works well in automation, because it is pretty clear what goals both can solve and what not. Is there some space between the two I didn't happen to run into as yet?

My interest in CSDP is more in CSDP itself, so psatz would be a side effect.

**Comment by palmskog** [comment]
One of the main benefits of CSDP is that we could potentially add [Validsdp](https://github.com/validsdp/validsdp) to the platform once we have it (assuming @proux01 and @erikmd are interested in this).

**Comment by MSoegtropIMC** [comment]
Indeed Validsdp is more what I am interested in. I have a lot of issues with LP solvers producing garbage results.

**Comment by MSoegtropIMC** [comment]
Triage note: this is still complicated to support on all platforms - help welcome!

**Comment by proux01** [comment]
A student advised by @erikmd recently added a csdp package to the OCaml OPAM repo. This works on various Linux and mac distribs but still requires testing/adaptation for Windows: https://github.com/ocaml/opam-repository/pull/25721

**Comment by MSoegtropIMC** [comment]
@proux01 : thanks for the note - i will look into it again.

**Comment by erikmd** [comment]
Yes @MSoegtropIMC: for the record, the approach followed by my student was exactly your initial suggestion:

> the easiest path might be to create an opam "from sources" package for csdp

→ https://opam.ocaml.org/packages/csdp/

The CI logs are not available anymore, but as far as I can tell, all Linux distributions tested in opam CI (except oraclelinux) were fine (e.g., debian + ubuntu + fedora + centos + opensuse + arch) as well as freebsd + macos-homebrew.
Even if csdp was already packaged in just a few distros of these.

But testing/adapting the package for windows-cygwin is still missing.

**Comment by MSoegtropIMC** [comment]
@erikmd : thanks! I had a quick look at the opam package - it looks doable on cygwin.

Is there a good reason for using sed instead of patch, say many lines matching each sed patch?

**Comment by proux01** [comment]
Not real reason, it was easy to use sed, but if patch is preferable for any reason, we can certainly change.

**Comment by erikmd** [comment]
The various sed lines just aim to:

* fix an upstream bug by adding a missing line in the sources (can be easily replaced with a patch)
* ensure the Makefile is properly configured for various os and distribution (the required Makefile tweaks might be doable with patch, but much more tricky I guess, as there are multiple cases…)

**Comment by erikmd** [comment]
Yes I am aware of this feature of patch.

But again, I am not sure it is adapted for the second kind of substitution I was talking about:

> * ensure the Makefile is properly configured for various os and distributions (the required Makefile tweaks might be doable with patch, but much more tricky I guess, as there are multiple cases…)

because it is conditional patching, i.e., only for some os / distributions (which does not seem to be possible with opam)

and the hypothesis whether "the [sed] does not apply any more" is very unlikely, because csdp development seems to have stalled, so the current version is the reference one AFAICT

**Comment by proux01** [comment]
The OPAM build rule currently calls sed, it could as well call patch (with the same conditions)

**Comment by erikmd** [comment]
OK! I better understand your idea.

@MSoegtropIMC is it "required" from your viewpoint to do this "refactoring"?

AFAIAC I won't have the time to do this in the upcoming weeks.

**Comment by MSoegtropIMC** [comment]
Indeed the idea is to have 4 patch files - one for each OS. This is for sure more readable and more error resilient.

As a compromise let's say it shall be changed in case there ever is a new version.

As Coq Platform maintainer I am having my share of updating other peoples opam packages, so I guess I am more picky about the maintainability of opam packages than others.

**Comment by MSoegtropIMC** [comment]
Maybe I change it in case I need to change something for cygwin.

**Comment by MSoegtropIMC** [comment]
@proux01 : we are currently discussing this in the board. First sorry for the long delays - it is still technically difficult to make a reliable multi-platform build of CSDP. We wonder if this is still needed by non linear arithmetic tactics, or if this request is meanwhile outdated.

**Comment by MSoegtropIMC** [comment]
CC Frérdéric Besson @fajb

**Comment by MSoegtropIMC** [comment]
Since the request is > 4 years old it just makes sense to ask the experts and re-assess this before making a decision.

The dependency of VaildSDP is noted!

An important question is if `psatz` still depends on it (`psatz` is mentioned in the 8.20 ref manual) and if `psatz` is still a recommended / actively maintained tactic or if there are plans to deprecate it (I don't suggest this in any way - I just ask what the plans are so that we have all information on the table).

Oh, I just saw that the 8.20 ref manual mentions in the header of https://coq.inria.fr/doc/V8.20.0/refman/addendum/micromega.html that `psatz` depends on `csdp`, so the first part of the above question is answered (this is not mentioned in the section of `psatz` as such - it probably should).

So only the second part of the question remains.

**Comment by proux01** [comment]
For the second part: we still have in the test-suite examples that can be solved only by psatz. So I'd say it still has some interest.

**Comment by MSoegtropIMC** [comment]
OK, thanks.

I had a look at the csdp opam package - looks good to me since it builds from sources. For not so frequently used Linux packages I am not a big fan of system supplied packages, since the versions one gets various too much.
